lib/PartitionedProducerImpl.cc

```diff
@@ -266,13 +266,10 @@ void PartitionedProducerImpl::closeAsync(CloseCallback originalCallback) {
             originalCallback(result);
         }
     };
-    if (state_ == Closed) {
-        closeCallback(ResultAlreadyClosed);
-        return;
-    }
-    State expectedState = Ready;
-    if (!state_.compare_exchange_strong(expectedState, Closing)) {
-        return;
+
+    State previous_state = state_.exchange(Closed);
+    if (previous_state == Closed) {
+        return closeCallback(ResultAlreadyClosed);
     }
```

Comment:
```suggestion
    State previousState = state_.exchange(Closed);
    if (previousState == Closed) {
        return closeCallback(ResultAlreadyClosed);
    }
```

Use camel case instead of snake case to make code style consistent. Or just simplify it to `if (state_.exchange(Closed) == Closed)`.

Reply:
In addition, we should follow the behavior of the Java client to set the state to `Closing`. Only after all sub-producers are closed should we set the state to `Closed`.

```java
        if (getState() == State.Closing || getState() == State.Closed) {
            return CompletableFuture.completedFuture(null);
        }
        setState(State.Closing); // here
```